clab/config/functions.go

```diff
@@ -0,0 +1,230 @@
+	// 		return nil, errors.New("required value not set")
+	// 	}
+	// 	return val, nil
+	// },
+	"ip": func(val interface{}) (interface{}, error) {
+		s := fmt.Sprintf("%v", val)
+		a := strings.Split(s, "/")
+		return a[0], nil
+	},
+	"ipmask": func(val interface{}) (interface{}, error) {
+		s := fmt.Sprintf("%v", val)
+		a := strings.Split(s, "/")
+		return a[1], nil
+	},
+	"default": func(in ...interface{}) (interface{}, error) {
+		if len(in) < 2 {
+			return nil, fmt.Errorf("default value expected")
+		}
+		if len(in) > 2 {
+			return nil, fmt.Errorf("too many arguments")
+		}
+
+		val := in[len(in)-1]
+		def := in[0]
+
+		switch v := val.(type) {
+		case nil:
+			return def, nil
+		case string:
+			if v == "" {
+				return def, nil
+			}
+		case bool:
+			if !v {
+				return def, nil
+			}
+		}
+		// if val == nil {
+		// 	return def, nil
+		// }
+
+		// If we have a input value, do some type checking
+		tval, tdef := typeof(val), typeof(def)
+		if tval == "string" && tdef == "int" {
+			if _, err := strconv.Atoi(val.(string)); err == nil {
+				tval = "int"
+			}
+			if tdef == "str" {
+				if _, err := strconv.Atoi(def.(string)); err == nil {
+					tdef = "int"
+				}
+			}
+		}
+		if tdef != tval {
+			return val, fmt.Errorf("expected type %v, got %v (value=%v)", tdef, tval, val)
+		}
+
+		// Return the value
+		return val, nil
+	},
+	"contains": func(substr string, str string) (interface{}, error) {
+		return strings.Contains(fmt.Sprintf("%v", str), fmt.Sprintf("%v", substr)), nil
+	},
+	"split": func(sep string, val interface{}) (interface{}, error) {
+		// Start and end values
+		if val == nil {
+			return []interface{}{}, nil
+		}
+		if sep == "" {
+			sep = " "
+		}
+
+		v := fmt.Sprintf("%v", val)
+
+		res := strings.Split(v, sep)
+		r := make([]interface{}, len(res))
+		for i, p := range res {
+			r[i] = p
+		}
+		return r, nil
+	},
+	"join": func(sep string, val interface{}) (interface{}, error) {
+		if sep == "" {
+			sep = " "
+		}
+		// Start and end values
+		switch v := val.(type) {
+		case []interface{}:
+			if val == nil {
+				return "", nil
+			}
+			res := make([]string, len(v))
+			for i, v := range v {
+				res[i] = fmt.Sprintf("%v", v)
+			}
+			return strings.Join(res, sep), nil
+		case []string:
+			return strings.Join(v, sep), nil
+		case []int, []int16, []int32:
+			return strings.Trim(strings.ReplaceAll(fmt.Sprint(v), " ", sep), "[]"), nil
+		}
+		return nil, fmt.Errorf("expected array [], got %v", val)
+	},
+	"slice": func(start, end int, val interface{}) (interface{}, error) {
+		// string or array
+		switch v := val.(type) {
+		case string:
+			if start < 0 {
+				start += len(v)
+			}
+			if end < 0 {
+				end += len(v)
+			}
+			return v[start:end], nil
+		case []interface{}:
+			if start < 0 {
+				start += len(v)
+			}
+			if end < 0 {
+				end += len(v)
+			}
+			return v[start:end], nil
+		}
+		return nil, fmt.Errorf("not an array")
+	},
+	"index": func(idx int, val interface{}) (interface{}, error) {
```

Comment:
since `template` package has an `index` builtin function, maybe you need to rename this one to something else?
https://pkg.go.dev/text/template#hdr-Functions

Reply:
See the comment on slice. It seems as if they have their parameters does not work with pipelining... or I'm missing something?

Reply:
gotcha, I left a comment there https://github.com/srl-labs/containerlab/pull/383#discussion_r634138965